osquery/utils/fgets_buffer.h

```diff
@@ -0,0 +1,205 @@
+      : NonblockingFile(),
+        path_(filepath),
+        selectTimeoutUsec_(selectTimeoutUsec) {
+    fd_ = ::open(path_.c_str(), O_RDONLY | O_NONBLOCK);
+  }
+
+  virtual ~NonblockingFileImpl() {}
+
+  /**
+   * If unable to open file in constructor, or
+   * close() has been called, this will return false.
+   */
+  bool isValid() override {
+    return fd_ > 0;
+  }
+
+  /**
+   * @return true if select() shows data is available for reading.
+   */
+  bool isDataAvail() override {
+    fd_set set;
+    struct timeval timeout = {0, (int)selectTimeoutUsec_};
+
+    FD_ZERO(&set); /* clear the set */
+    FD_SET(fd_, &set); /* add our file descriptor to the set */
+
+    int rv = select(fd_ + 1, &set, NULL, NULL, &timeout);
+    if (rv == -1) {
+      // TODO : after a certain number of these, shut it down?
+    }
+    return rv > 0;
+  }
+
+  /**
+   * Reads data from file, appending to buf upto capacity.
+   * @return number of bytes read, -1 on error, 0 on none.
+   */
+  int64_t nbRead(std::vector<char>& buf) override {
+    char tmpbuf[4096];
+    size_t remaining = buf.capacity() - buf.size();
+    auto len = (remaining > sizeof(tmpbuf) ? sizeof(tmpbuf) : remaining);
+    int64_t bytesRead = ::read(fd_, tmpbuf, len);
+
+    if (bytesRead > 0) {
+      auto p = buf.data() + buf.size();
+      buf.resize(buf.size() + bytesRead);
+      memcpy(p, tmpbuf, bytesRead);
+    }
+    return bytesRead;
+  }
+
+  void close() override {
+    if (fd_ > 0) {
+      ::close(fd_);
+      fd_ = 0;
+    }
+  }
+
+ protected:
+  std::string path_;
+  int fd_{-1};
+  uint32_t selectTimeoutUsec_;
+};
+#endif // OSQUERY_POSIX
+
+class FgetsBuffer {
+ public:
+  /**
+   * If includeNewline is true, then strings returned by fgets() will
+   * have newline character at the end (like stdio behavior).
+   */
+  FgetsBuffer(SPNonblockingFile spFile,
+              size_t maxLineLen = 16384,
+              bool includeNewline = false)
+      : spFile_(spFile),
+        buf_(),
+        maxLineLen_(maxLineLen),
+        includeNewline_(includeNewline) {
+    buf_.reserve(maxLineLen);
+  }
+
+  /**
+   * If buffer reaches maxLineLen without a newline,
+   * buffer is cleared.  This counter returns number of
+   * chars dropped.
+   */
+  size_t getNumDroppedChars() {
+    return droppedChars_;
+  }
+
+  /// return false on success getting a line into dest, true on timeout or error
+  bool fgets(std::string& dest) {
+    if (!spFile_->isValid()) {
+      return true;
+    }
+
+    // first see if a line is already buffered
+
+    if (!buf_.empty() && !_gets(dest)) {
+      return false;
+    }
+
+    if (spFile_->isDataAvail()) {
+      // read more
+
+      int64_t bytesRead = spFile_->nbRead(buf_);
+      if (bytesRead > 0) {
+        // try again to see if entire line is available
+
+        return _gets(dest);
+      }
+    }
+    return true;
+  }
+
+  SPNonblockingFile getFile() {
+    return spFile_;
```

Comment:
Given the discussion about ownership, is this really needed?.
As far as I see it's not used, and when changing everything to use a `unique_ptr`, this should return a raw pointer.
The question though is, why we need to return a modifiable pointer to guts of FgetsBuffer; why should we give access to it at all?

Reply:
Yes, nice catch.  It was originally part of interface and removed, but not from this implementation.